File: python/best_time_to_buy_and_sell_stock.py
from typing import List

# Use a modified version of Kadane's max sub array algorithm


def maxProfit(prices: List[int]) -> int:
    lmax, gmax = 0, 0
    for i in range(1, len(prices)):
        lmax = max(0, lmax + prices[i] - prices[i - 1]) #I can only buy and sell once
        gmax = max(lmax, gmax)
    return gmax

File: cpp/best_time_to_buy_and_sell_stock.cpp
#include <vector>

using namespace std;

class Solution {
public:
    //Kadane's algorithm
    int max (int a, int b) {
        return a > b ? a : b;
    }
    
    int maxProfit(vector<int>& prices) {
        int lsum = 0, gsum = 0;
        for (int i = 1; i < prices.size(); ++i) {
            lsum = max(0, lsum + prices[i] - prices[i - 1]);
            gsum = max(lsum, gsum);
        }
        return gsum;
    }
};